File: Python/Two Pointers/M-438. Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        needed_chars = {}
        required_length = len(p)
        result = []

        for c in p:
            needed_chars[c] = (needed_chars.get(c) or 0) + 1

        l, r = 0, 0
        while r < len(s):
            if s[r] in needed_chars:
                if needed_chars[s[r]] > 0:
                    required_length -= 1
                needed_chars[s[r]] -= 1

            if r - l == len(p) - 1:
                if required_length == 0:
                    result.append(l)

                if s[l] in needed_chars:
                    if needed_chars[s[l]] >= 0:
                        required_length += 1
                    needed_chars[s[l]] += 1
                l += 1

            r += 1

        return result

File: Javascript/Two Pointers/M-438. Find All Anagrams in a String.ts
function findAnagrams(s: string, p: string): number[] {
  const result: number[] = [];
  const neededChars = {};
  let requiredLength = p.length;
  for (const c of p) neededChars[c] = (neededChars[c] || 0) + 1;

  let l = 0;
  let r = 0;
  while (r < s.length) {
    if (neededChars[s[r]] > 0) requiredLength--;
    neededChars[s[r]]--;

    /* reach to window size */
    if (r - l === p.length - 1) {
      if (requiredLength === 0) result.push(l);
      if (neededChars[s[l]] >= 0) requiredLength++;
      neededChars[s[l]]++;
      l++;
    }
    r++;
  }

  return result;
}

const testData = [
  {
    s: 'cbaebabacd',
    p: 'abc',
    ans: [0, 6],
  },
  {
    s: 'abab',
    p: 'ab',
    ans: [0, 1, 2],
  },
];

for (const data of testData) {
  console.log(
    JSON.stringify(data.ans) === JSON.stringify(findAnagrams(data.s, data.p))
  );
}
